pennylane_lightning_gpu/src/simulator/StateVectorCudaMPI.hpp

```diff
@@ -0,0 +1,1455 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+/**
+ * @file StateVectorCudaMPI.hpp
+ */
+#pragma once
+
+#include <functional>
+#include <numeric>
+#include <random>
+#include <unordered_map>
+#include <unordered_set>
+#include <vector>
+
+#include <cuComplex.h> // cuDoubleComplex
+#include <cuda.h>
+#include <custatevec.h> // custatevecApplyMatrix
+
+#include "Constant.hpp"
+#include "Error.hpp"
+#include "MPIManager.hpp"
+#include "MPIWorker.hpp"
+#include "StateVectorCudaBase.hpp"
+#include "cuGateCache.hpp"
+#include "cuGates_host.hpp"
+#include "cuda_helpers.hpp"
+
+/// @cond DEV
+namespace {
+namespace cuUtil = Pennylane::CUDA::Util;
+using namespace Pennylane::CUDA;
+using namespace Pennylane::MPI;
+using namespace Pennylane::Util;
+} // namespace
+/// @endcond
+
+namespace Pennylane {
+
+// declarations of external functions (defined in initSV.cu).
+extern void setStateVector_CUDA(cuComplex *sv, int &num_indices,
+                                cuComplex *value, int *indices,
+                                size_t thread_per_block,
+                                cudaStream_t stream_id);
+extern void setStateVector_CUDA(cuDoubleComplex *sv, long &num_indices,
+                                cuDoubleComplex *value, long *indices,
+                                size_t thread_per_block,
+                                cudaStream_t stream_id);
+
+extern void setBasisState_CUDA(cuComplex *sv, cuComplex &value,
+                               const size_t index, bool async,
+                               cudaStream_t stream_id);
+extern void setBasisState_CUDA(cuDoubleComplex *sv, cuDoubleComplex &value,
+                               const size_t index, bool async,
+                               cudaStream_t stream_id);
+
+/**
+ * @brief Managed memory CUDA state-vector class using custateVec backed
+ * gate-calls.
+ *
+ * @tparam Precision Floating-point precision type.
+ */
+template <class Precision>
+class StateVectorCudaMPI
+    : public StateVectorCudaBase<Precision, StateVectorCudaMPI<Precision>> {
+  private:
+    using BaseType = StateVectorCudaBase<Precision, StateVectorCudaMPI>;
+
+    size_t numGlobalQubits_;
+    size_t numLocalQubits_;
+    MPIManager mpi_manager_;
+
+    SharedCusvHandle handle_;
+    SharedCublasCaller cublascaller_;
+    mutable SharedCusparseHandle
+        cusparsehandle_; // This member is mutable to allow lazy initialization.
+    SharedLocalStream localStream_;
+    SharedMPIWorker svSegSwapWorker_;
+    GateCache<Precision> gate_cache_;
+
+  public:
+    using CFP_t =
+        typename StateVectorCudaBase<Precision,
+                                     StateVectorCudaMPI<Precision>>::CFP_t;
+    using GateType = CFP_t *;
+
+    StateVectorCudaMPI() = delete;
+
+    StateVectorCudaMPI(MPIManager mpi_manager, size_t mpi_buffer_size,
```

Comment:
Would it make sense to make `mpi_buffer_size` and attribute of `MPIManager` objects?

Reply:
I would prefer to keep as it is for the moment since `mpi_buffer_size` is only required for creating `custatevecSVSwapWorkerDescriptor`.